src/Patch/Document.js

```diff
@@ -0,0 +1,67 @@
+import Native from './Native';
+import CustomElementInternals from '../CustomElementInternals';
+import * as Utilities from '../Utilities';
+
+import PatchParentNode from './Interface/ParentNode';
+
+/**
+ * @param {!CustomElementInternals} internals
+ */
+export default function(internals) {
+  Utilities.setPropertyUnchecked(Document.prototype, 'createElement',
+    /**
+     * @this {Document}
+     * @param {string} localName
+     * @return {!Element}
+     */
+    function(localName) {
+      const definition = internals.localNameToDefinition(localName);
+      if (definition) {
+        return new (definition.constructor)();
+      }
+
+      const result = /** @type {!Element} */
+        (Native.Document_createElement.call(this, localName));
+      internals.patch(result);
+      return result;
+    });
+
+  Utilities.setPropertyUnchecked(Document.prototype, 'importNode',
+    /**
+     * @this {Document}
+     * @param {!Node} node
+     * @param {boolean=} deep
+     * @return {!Node}
+     */
+    function(node, deep) {
+      const clone = Native.Document_importNode.call(this, node, deep);
+      internals.patchTree(clone);
+      internals.upgradeTree(clone);
```

Comment:
Upgrades should only happen in the context of documents with an associated registry now. Still need to avoid the double walk.